### DISCUSSION TITLE
Socratic Reasoning and Dialogue

### DISCUSSION THREAD

**Discussion by BradKML** [SmartManoj/Kevin#145] [Ideas]
Thinking how knowledge self-accreditation can happen with models or data. Could be useful
Some example repo to start a discussion https://github.com/RunzheYang/SocraticAI
1. Can Socratic Reasoning be used to fine-tune models? Or teach X-of-Thought reasoning?
2. Can the logs of Socratic Reasoning be used as the data part of a RAG?
3. Can multiple models argue and contest one another for the best solutions?
4. What is with Socratic Reasoning and resisting model collapse? https://github.com/Yiwen-Ding/Guided-Self-Improvement https://arxiv.org/html/2402.07043v2
5. What other LLM as Judge models are there? e.g. https://github.com/SmartManoj/Kevin/discussions/136#discussioncomment-11615086

**Comment by BradKML** [comment]
Tutoring might be Socratic-adjacent at least so I will need to keep an eye on that https://github.com/JushBJJ/Mr.-Ranedeer-AI-Tutor

**Comment by SmartManoj** [comment]
Which models are you looking to fine-tune? How does that model perform on [these prompts](https://medium.com/@srinivas.jayanthi/prompting-strategy-reasoning-109a55d67d16)?

**Reply by BradKML** [reply]
The major issue is that Socratic reasoning or dialectical reasoning, is that the two agents has to take on somewhat opposing beliefs, rather than one coherent role without self-doubt. Hop between questions, examples, disagreements, and clarifications.

**Reply by SmartManoj** [reply]
Then the conversation will never end. 

https://www.kaggle.com/competitions/llms-you-cant-please-them-all

**Reply by BradKML** [reply]
Notice I say somewhat opposing beliefs, not totally opposing. And it "never ending" metaphorically means that it can keep exploring an idea until all the major valid arguments can be exhausted, throwing in a moderator agent to limit obvious things that need not be said. The arguing agents would be forbidden to self-contradict as well, and will concede when they change their mind or stuck in a loop of failed rationalization. They can spawn a different agent that "moves the goal post" (have a slightly different stance that is more valid than the previous one), which makes meeting in the middle easier.

**Comment by SmartManoj** [comment]
What do you think about [Instruction Following Eval](https://github.com/google-research/google-research/tree/master/instruction_following_eval) ([HF Link](https://huggingface.co/datasets/google/IFEval))?

**Reply by BradKML** [reply]
Needed but hard to evaluate, also there are models that are more dedicated to tools than code, which is odd cus why not do multi-objective optimization?

**Reply by SmartManoj** [reply]
Easy to evaluate within ~5 mins. The evaluation step is [given](https://github.com/google-research/google-research/tree/master/instruction_following_eval#how-to-run) in Github.

One of the prompts:
Write a logic quiz for teenagers about a chesterfield. In your entire response, the letter t should appear at most once.

**Reply by BradKML** [reply]
Oh yeah maybe there is a difference between complex instruction following vs simple ones? Might need to check in that case. But a moderator agent built based on instruction following evals would be really useful.
